robot/client/client.go

```diff
@@ -897,29 +900,32 @@ func (rc *RobotClient) StopAll(ctx context.Context, extra map[resource.Name]map[
 
 // Log sends a log entry to the server. To be used by Golang modules wanting to
 // log over gRPC and not by normal Golang SDK clients.
-func (rc *RobotClient) Log(ctx context.Context, log zapcore.Entry, fields []zapcore.Field) error {
-	// TODO(RSDK-6280): Preserve the type of all `fields`.
+func (rc *RobotClient) Log(ctx context.Context, log zapcore.Entry, fields []zap.Field) error {
 	message := fmt.Sprintf("%v\t%v", log.Caller.TrimmedPath(), log.Message)
-	for i, field := range fields {
-		if i == 0 {
-			message = fmt.Sprintf("%v\t{%q: %q", message, field.Key, field.String)
-		} else {
-			message = fmt.Sprintf("%v, %q: %q", message, field.Key, field.String)
+
+	fieldsP := make([]*structpb.Struct, 0, len(fields))
+	for _, field := range fields {
+		fieldP, err := logging.FieldToProto(field)
+		if err != nil {
+			return err
 		}
+		fieldsP = append(fieldsP, fieldP)
 	}
-	message = fmt.Sprintf("%v}", message) // close }
 
 	logRequest := &pb.LogRequest{
-		// no batching for now (one LogEntry at a time).
+		// No batching for now (one LogEntry at a time).
 		Logs: []*commonpb.LogEntry{{
-			// leave out Host; Host is not currently meaningful
-			Level: log.Level.String(),
-			// leave out Time; Time is already in Message field below
+			// Leave out Host; Host is not currently meaningful.
+			Level:      log.Level.String(),
+			Time:       timestamppb.New(time.Now()),
```

Comment:
Probably just an oversight -- I wasn't expecting to see a `time.Now()` call here.

The input `log` zapcore.Entry has a `Time` [object](https://github.com/uber-go/zap/blob/8f5ee80ab2dbc713823341ce30334cd9c03a98e5/zapcore/entry.go#L145) on it [that we've (presumably) set](https://github.com/viamrobotics/rdk/blob/5ce3ca953af36fcf09ce888bafa9709166d58c98/logging/impl.go#L39) already. I'd expect this code to use that.

Reply:
> The input log zapcore.Entry has a Time [object](https://github.com/uber-go/zap/blob/8f5ee80ab2dbc713823341ce30334cd9c03a98e5/zapcore/entry.go#L145) on it [that we've (presumably) set](https://github.com/viamrobotics/rdk/blob/5ce3ca953af36fcf09ce888bafa9709166d58c98/logging/impl.go#L39) already. I'd expect this code to use that.

Yep! My oversight and good catch on your part. Switched to `log.Time`.